### PR TITLE
Fix scalar leak in WindowRetrySuite

### LINKED ISSUE #10499: [BUG] Unit tests core dump as below
**Describe the bug**

Unit tests core dump on `branch-24.04` as below, JDK11-nightly/545
[hs_err_pid2664.log](https://github.com/NVIDIA/spark-rapids/files/14396678/hs_err_pid2664.log)

```
 - row based group by window handles GpuRetryOOM
 - row-based group by running window handles GpuSplitAndRetryOOM
 #
 # A fatal error has been detected by the Java Runtime Environment:
 #
 #  SIGSEGV (0xb) at pc=0x0000000000000000, pid=2664, tid=2755
 #
 # JRE version: OpenJDK Runtime Environment (11.0.21+9) (build 11.0.21+9-post-Ubuntu-0ubuntu120.04)
 # Java VM: OpenJDK 64-Bit Server VM (11.0.21+9-post-Ubuntu-0ubuntu120.04, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
 # Problematic frame:
 # C  0x0000000000000000
 #
 # Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/core.2664)
 #
 # An error report file with more information is saved as:
 # /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/hs_err_pid2664.log
 Compiled method (c1) 1191062 28589  s!   3       ai.rapids.cudf.Scalar$OffHeapState::cleanImpl (133 bytes)
  total in heap  [0x00007f78c1f44910,0x00007f78c1f46678] = 7528
  relocation     [0x00007f78c1f44a88,0x00007f78c1f44c60] = 472
  main code      [0x00007f78c1f44c60,0x00007f78c1f45d60] = 4352
  stub code      [0x00007f78c1f45d60,0x00007f78c1f45e80] = 288
  oops           [0x00007f78c1f45e80,0x00007f78c1f45ea8] = 40
  metadata       [0x00007f78c1f45ea8,0x00007f78c1f45f28] = 128
  scopes data    [0x00007f78c1f45f28,0x00007f78c1f461c0] = 664
  scopes pcs     [0x00007f78c1f461c0,0x00007f78c1f46560] = 928
  dependencies   [0x00007f78c1f46560,0x00007f78c1f46580] = 32
  handler table  [0x00007f78c1f46580,0x00007f78c1f46610] = 144
  nul chk table  [0x00007f78c1f46610,0x00007f78c1f46678] = 104
 Compiled method (c1) 1191062 28589  s!   3       ai.rapids.cudf.Scalar$OffHeapState::cleanImpl (133 bytes)
  total in heap  [0x00007f78c1f44910,0x00007f78c1f46678] = 7528
  relocation     [0x00007f78c1f44a88,0x00007f78c1f44c60] = 472
  main code      [0x00007f78c1f44c60,0x00007f78c1f45d60] = 4352
  stub code      [0x00007f78c1f45d60,0x00007f78c1f45e80] = 288
  oops           [0x00007f78c1f45e80,0x00007f78c1f45ea8] = 40
  metadata       [0x00007f78c1f45ea8,0x00007f78c1f45f28] = 128
  scopes data    [0x00007f78c1f45f28,0x00007f78c1f461c0] = 664
  scopes pcs     [0x00007f78c1f461c0,0x00007f78c1f46560] = 928
  dependencies   [0x00007f78c1f46560,0x00007f78c1f46580] = 32
  handler table  [0x00007f78c1f46580,0x00007f78c1f46610] = 144
  nul chk table  [0x00007f78c1f46610,0x00007f78c1f46678] = 104
 #
 # If you would like to submit a bug report, please visit:
 #   https://bugs.launchpad.net/ubuntu/+source/openjdk-lts
 # The crash happened outside the Java Virtual Machine in native code.
 # See problematic frame for where to report the bug.
 #
 [INFO] ------------------------------------------------------------------------
 [INFO] Reactor Summary for RAPIDS Accelerator for Apache Spark Root Project 24.04.0-SNAPSHOT:
 [INFO] 
 [INFO] RAPIDS Accelerator for Apache Spark Root Project ... SUCCESS [ 15.341 s]
 [INFO] rapids-4-spark-jdk-profiles_2.12 ................... SUCCESS [  0.311 s]
 [INFO] rapids-4-spark-shim-deps-parent_2.12 ............... SUCCESS [ 11.861 s]
 [INFO] rapids-4-spark-sql-plugin-api_2.12 ................. SUCCESS [ 24.701 s]
 [INFO] RAPIDS Accelerator for Apache Spark SQL Plugin ..... SUCCESS [01:23 min]
 [INFO] RAPIDS Accelerator for Apache Spark Shuffle Plugin . SUCCESS [  7.602 s]
 [INFO] RAPIDS Accelerator for Apache Spark Scala UDF Plugin SUCCESS [ 39.216 s]
 [INFO] RAPIDS Accelerator for Apache Spark Delta Lake 2.0.x Support SUCCESS [ 13.406 s]
 [INFO] RAPIDS Accelerator for Apache Spark Aggregator ..... SUCCESS [  9.863 s]
 [INFO] Data Generator ..................................... SUCCESS [ 10.006 s]
 [INFO] RAPIDS Accelerator for Apache Spark Distribution ... SUCCESS [01:24 min]
 [INFO] rapids-4-spark-integration-tests_2.12 .............. SUCCESS [ 36.575 s]
 [INFO] RAPIDS Accelerator for Apache Spark Tests .......... FAILURE [20:26 min]
 [INFO] RAPIDS Accelerator for Apache Spark Tools Support .. SKIPPED
 [INFO] ------------------------------------------------------------------------
 [INFO] BUILD FAILURE
 [INFO] ------------------------------------------------------------------------
 [INFO] Total time:  26:03 min
 [INFO] Finished at: 2024-02-25T02:27:11Z
 [INFO] ------------------------------------------------------------------------
 [ERROR] Failed to execute goal org.scalatest:scalatest-maven-plugin:2.0.2:test (test) on project rapids-4-spark-tests_2.12: There are test failures -> 



-------------------------------------------------------------------------------------



--------------  S U M M A R Y ------------

Command Line: -Dai.rapids.refcount.debug=true -Djava.awt.headless=true -Djava.io.tmpdir=/home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/tmp -Drapids.shuffle.manager.override=true -Dspark.ui.enabled=false -Dspark.ui.showConsoleProgress=false -Dspark.unsafe.exceptionOnMemoryLeak=true -Dbasedir=/home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests -javaagent:/root/.m2/repository/org/jacoco/org.jacoco.agent/0.8.8/org.jacoco.agent-0.8.8-runtime.jar=destfile=/home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/jacoco.exec,append=true,includes=ai.rapids.cudf.*:com.nvidia.spark.*:org.apache.spark.sql.rapids.*,excludes=spark320.com.nvidia.shaded.spark.* -ea -Xmx4g -Xss4m -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED -Djdk.reflect.useDirectMethodHandle=false org.scalatest.tools.Runner -R /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/classes /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/test-classes -o -f /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/surefire-reports/scala-test-output.txt -u /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/surefire-reports/.

Host: AMD EPYC 7313P 16-Core Processor, 32 cores, 60G, Ubuntu 20.04.6 LTS
Time: Sun Feb 25 02:27:10 2024 UTC elapsed time: 1191.061316 seconds (0d 0h 19m 51s)

---------------  T H R E A D  ---------------

Current thread (0x00007f77e8352000):  JavaThread "Cleaner Thread" daemon [_thread_in_native, id=2755, stack(0x00007f780c3f7000,0x00007f780c7f8000)]

Stack: [0x00007f780c3f7000,0x00007f780c7f8000],  sp=0x00007f780c7f6718,  free space=4093k
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 28591  ai.rapids.cudf.Scalar.closeScalar(J)V (0 bytes) @ 0x00007f78c8c06a11 [0x00007f78c8c069c0+0x0000000000000051]
J 28589 c1 ai.rapids.cudf.Scalar$OffHeapState.cleanImpl(Z)Z (133 bytes) @ 0x00007f78c1f457d4 [0x00007f78c1f44c60+0x0000000000000b74]
J 61839 c2 ai.rapids.cudf.MemoryCleaner$CleanerWeakReference.clean()V (42 bytes) @ 0x00007f78cb1d7aa8 [0x00007f78cb1d79e0+0x00000000000000c8]
j  ai.rapids.cudf.MemoryCleaner.lambda$static$0()V+126
j  ai.rapids.cudf.MemoryCleaner$$Lambda$2057.run()V+0
j  java.lang.Thread.run()V+11 java.base@11.0.21
v  ~StubRoutines::call_stub

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000000

Register to memory mapping:

RAX=0x0 is NULL
RBX={method} {0x00007f7898b3b238} 'closeScalar' '(J)V' in 'ai/rapids/cudf/Scalar'
RCX=0x0000000000000002 is an unknown value
RDX=0x0000000000000010 is an unknown value
RSP=0x00007f780c7f6718 is pointing into the stack for thread: 0x00007f77e8352000
RBP=0x00007f778e8c3b10 points into unknown readable memory: 0x00007f77e4daeae8 | e8 ea da e4 77 7f 00 00
RSI=0x00007f740e201400 is an unknown value
RDI=0x00007f73fa3c6d60 points into unknown readable memory: 0x00007f73fad5d170 | 70 d1 d5 fa 73 7f 00 00
R8 =0x0000000000000002 is an unknown value
R9 =0x00007f77e83529f0 points into unknown readable memory: 0x0000000000000006 | 06 00 00 00 00 00 00 00
R10=0x00007f77c04768e0: Java_ai_rapids_cudf_Scalar_closeScalar+0x0000000000000000 in /home/jenkins/agent/workspace/jenkins-JDK11-nightly-545/tests/target/spark320/tmp/cudf5094709651714106938.so at 0x00007f77bf6bc000
R11=0x00007f78db090a70 points into unknown readable memory: 0x0000000000000000 | 00 00 00 00 00 00 00 00
R12=0x0 is NULL
R13=0x00007f780c7f672c is pointing into the stack for thread: 0x00007f77e8352000
R14=0x0 is NULL
R15=0x00007f77e8352000 is a thread

```

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #10499.  WindowRetrySuite leaks because it does not close the GpuRunningWindowIterator which internally has a Scalar instance.  Updated GpuRunningWindowIterator to derive from GpuColumnarBatchIterator to reduce some boilerplate code and updated the test to always close the iterator instance to avoid the leak.

### CODE REVIEW

**Review by jbrennan333 — APPROVED** [review]
Looks good!

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuRunningWindowExec.scala [hunk 1]
```diff
@@ -75,15 +70,12 @@ class GpuRunningWindowIterator(
...
-    if (!isClosed) {
-      isClosed = true
-      fixerIndexMap.values.foreach(_.close())
-      saveLastParts(Array.empty)
-      saveLastOrder(Array.empty)
-      cachedBatch.foreach(_.close())
-      cachedBatch = None
-    }
+  override def doClose(): Unit = {
+    fixerIndexMap.values.foreach(_.close())
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuRunningWindowExec.scala:74]: should we fix `foreach(_.close())` in this file to call safeClose in this PR or make it a standalone PR?

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuRunningWindowExec.scala:74]: Since premerge is blocked, moving this to a followup.

**Review by gerashegalov — APPROVED** [review]
LGTM, might need a separate issue promoting safeClose

**Review by revans2 — APPROVED** [review]

**Pull request merged**